src/Index.hpp

```diff
@@ -110,6 +110,131 @@ namespace nbs {
             return types;
         }
 
+
+        // Get the timestamp where as many subtypes move forward without moving forward twice (still allowed when passed
+        // a single type)
+        std::uint64_t nextTimestamp(const uint64_t& timestamp,
+                                    const std::vector<TypeSubtype>& types,
+                                    const int& steps) {
+
+            auto matchingIndexItems = this->getIteratorsForTypes(types);
+
+            std::vector<TimestampState> states;
+
+            for (auto& iteratorPair : matchingIndexItems) {
+                auto& begin = iteratorPair.first;
+                auto& end   = iteratorPair.second;
+
+                IndexItemFile target;
+                target.item.timestamp = timestamp;
+
+
+                if (begin != end) {
+                    // Find the first item with a timestamp greater than the requested timestamp
+                    auto position =
+                        std::upper_bound(begin, end, target, [](const IndexItemFile& a, const IndexItemFile& b) {
+                            return a.item.timestamp < b.item.timestamp;
+                        });
+
+                    // This is before the beginning, the first packet is actually begin
+                    if (position != begin || position != end) {
+                        position = std::prev(position, 1);
+                    }
+                    else {
+                        throw std::invalid_argument("No packet found for this type at the requested timestamp");
+                    }
+
+                    states.push_back(TimestampState{0, position, begin, end});
+                }
+            }
+
+            if (states.empty()) {
+                throw std::runtime_error("File with requested types exist but contain no messages. ");
+            }
+
+            // Traverse through input type(s)
+            while (true) {
+                if (steps > 0) {
+                    auto min_val = std::min_element(  //
+                        states.begin(),
+                        states.end(),
+                        [](TimestampState& a, TimestampState& b) {
+                            auto a_time = std::next(a.position) == a.end ? std::numeric_limits<uint64_t>::max()
+                                                                         : std::next(a.position)->item.timestamp;
+                            auto b_time = std::next(b.position) == b.end ? std::numeric_limits<uint64_t>::max()
+                                                                         : std::next(b.position)->item.timestamp;
+                            return a_time < b_time;
+                        });
+
+                    // Check if the next step would move us off the end of the nbs file.
+                    if (std::distance(min_val->position, min_val->end) == 0) {
```

Comment:
You can just check if it is the end
```suggestion
                    if (min_val->position == min_val->end) {
```